protocols/upnp/src/behaviour.rs

```diff
@@ -396,27 +396,29 @@ where
         loop {
             match self.state {
                 GatewayState::Searching(ref mut fut) => match Pin::new(fut).poll(cx) {
-                    Poll::Ready(result) => match result {
-                        Ok(gateway) => {
-                            if !is_addr_global(gateway.external_addr) {
-                                self.state =
-                                    GatewayState::NonRoutableGateway(gateway.external_addr);
-                                log::debug!(
-                                    "the gateway is not routable, its address is {}",
-                                    gateway.external_addr
-                                );
-                                return Poll::Ready(ToSwarm::GenerateEvent(
-                                    Event::NonRoutableGateway,
-                                ));
+                    Poll::Ready(result) => {
+                        match result.expect("sender shouldn't have been dropped") {
+                            Ok(gateway) => {
+                                if !is_addr_global(gateway.external_addr) {
```

Comment:
What if we do this already in the `search` function and return an `Err` directly? You could send a "double" `Result` through the oneshot if you don't want to make a separate enum. See https://sled.rs/errors.

Reply:
I'd prefer it to be typed instead of strings but I'll leave it up to you :)